File: Algorithms/p200_Number_of_Islands/p200_Number_of_Islands_2.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        if(grid.empty()||grid[0].empty()) return 0;
        int m = grid.size();
        int n = grid[0].size();
        int num = 0;
        vector<vector<int>> visited(m,vector<int>(n,0));
        queue<pair<int,int>> q;
        for(int i=0;i<m;++i){
            for(int j=0;j<n;++j){
                if(grid[i][j]=='1'&&visited[i][j]==0){
                    q.push(make_pair(i,j));
                    visited[i][j] = 1;
                    while(!q.empty()){
                        int i1 = q.front().first;
                        int j1 = q.front().second;
                        q.pop();
                        if(i1>0&&visited[i1-1][j1]==0&&grid[i1-1][j1]=='1'){
                            visited[i1-1][j1] = 1;
                            q.emplace(i1-1,j1);
                        }
                        if(i1<m-1&&visited[i1+1][j1]==0&&grid[i1+1][j1]=='1'){
                            visited[i1+1][j1] = 1;
                            q.emplace(i1+1,j1);
                        }
                        if(j1>0&&visited[i1][j1-1]==0&&grid[i1][j1-1]=='1'){
                            visited[i1][j1-1] = 1;
                            q.emplace(i1,j1-1);
                        }
                        if(j1<n-1&&visited[i1][j1+1]==0&&grid[i1][j1+1]=='1'){
                            visited[i1][j1+1] = 1;
                            q.emplace(i1,j1+1);
                        }
                    }
                    ++num;
                }
            }
        }
        return num;
    }
};

File: Algorithms/p200_Number_of_Islands/p200_Number_of_Islands_2.py
class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        if not grid or not grid[0]:
            return 0
        m = len(grid)
        n = len(grid[0])
        num = 0
        visited = [[False for j in range(n)] for i in range(m)]
        q = collections.deque()
        for i in range(m):
            for j in range(n):
                if grid[i][j]=='1' and not visited[i][j]:
                    q.append((i,j))
                    visited[i][j] = True
                    while q:
                        i1,j1 = q.popleft()
                        if i1>0 and not visited[i1-1][j1] and grid[i1-1][j1]=='1':
                            visited[i1-1][j1] = True
                            q.append((i1-1,j1))
                        if i1<m-1 and not visited[i1+1][j1] and grid[i1+1][j1]=='1':
                            visited[i1+1][j1] = True
                            q.append((i1+1,j1))
                        if j1>0 and not visited[i1][j1-1] and grid[i1][j1-1]=='1':
                            visited[i1][j1-1] = True
                            q.append((i1,j1-1))
                        if j1<n-1 and not visited[i1][j1+1] and grid[i1][j1+1]=='1':
                            visited[i1][j1+1] = True
                            q.append((i1,j1+1))
                    num += 1
        return num
